storage/src/journal/fixed.rs

```diff
@@ -328,58 +358,69 @@ impl<E: Storage + Metrics, A: Codec<Cfg = ()> + FixedSize> Journal<E, A> {
     pub async fn oldest_retained_pos(&self) -> Result<Option<u64>, Error> {
-        let (oldest_blob_index, blob) = self.oldest_blob();
-        if blob.size().await == 0 {
+        let oldest_blob_index = self.oldest_blob_index();
+        if oldest_blob_index == self.tail_index && self.tail.size().await == 0 {
             return Ok(None);
         }
+
         // The oldest retained item is the first item in the oldest blob.
-        Ok(Some(*oldest_blob_index * self.cfg.items_per_blob))
+        Ok(Some(oldest_blob_index * self.cfg.items_per_blob))
     }
 
     /// Read the item at the given position in the journal.
     pub async fn read(&self, item_pos: u64) -> Result<A, Error> {
         let blob_index = item_pos / self.cfg.items_per_blob;
+        if blob_index > self.tail_index {
+            return Err(Error::InvalidItem(item_pos));
+        }
+        let item_index = item_pos % self.cfg.items_per_blob;
+        let offset = item_index * Self::CHUNK_SIZE_U64;
 
-        let blob = match self.blobs.get(&blob_index) {
-            Some(blob) => blob,
-            None => {
-                let (newest_blob_index, _) = self.newest_blob();
-                if blob_index > *newest_blob_index {
-                    return Err(Error::InvalidItem(item_pos));
-                }
-                let (oldest_blob_index, _) = self.oldest_blob();
-                assert!(blob_index < *oldest_blob_index);
-                return Err(Error::ItemPruned(item_pos));
-            }
+        if blob_index == self.tail_index {
+            let read = self
+                .tail
+                .read_at(vec![0u8; Self::CHUNK_SIZE], offset)
+                .await?;
+            return Self::verify_integrity(read.as_ref());
+        }
+
+        let Some(blob) = self.blobs.get(&blob_index) else {
+            return Err(Error::ItemPruned(item_pos));
         };
 
-        let item_index = item_pos % self.cfg.items_per_blob;
-        let offset = item_index * Self::CHUNK_SIZE_U64;
         let read = blob.read_at(vec![0u8; Self::CHUNK_SIZE], offset).await?;
         Self::verify_integrity(read.as_ref())
     }
 
-    /// Verify the integrity of the Array + checksum in `buf`, returning the array if it is valid.
+    /// Verify the integrity of the Array + checksum in `buf`, returning:
+    /// - The array if it is valid,
+    /// - Error::ChecksumMismatch if the checksum is invalid, or
+    /// - Error::Codec if the array could not be decoded after passing the checksum check.
+    ///
+    ///  Error::Codec likely indicates a logic error rather than a corruption issue.
     fn verify_integrity(buf: &[u8]) -> Result<A, Error> {
         let stored_checksum = u32::from_be_bytes(buf[A::SIZE..].try_into().unwrap());
         let checksum = crc32fast::hash(&buf[..A::SIZE]);
         if checksum != stored_checksum {
             return Err(Error::ChecksumMismatch(stored_checksum, checksum));
         }
-        Ok(A::decode(&buf[..A::SIZE]).unwrap())
```

Comment:
noticed this unwrap() here which didn't smell right.  Returning the CodeError explicitly now, even though it is probably an unrecoverable condition.